### PR TITLE
added string_add lint and fixed string_add_assign + test

### LINKED ISSUE #121: Warn on use of `+` for String
I honestly think that the `Add` implementation for String was a mistake – it isn't symmetric even in type signature, and it consumes self, which is surprising to the newbie (who probably has the mental model of other languages, where a new String would be created).

So we should warn the newbie of this pitfall, and maybe suggest using [`String::push_str`](http://doc.rust-lang.org/std/string/struct.String.html#method.push_str) instead (perhaps only on self-assignment, like `s = s + "..."`)

### PR DESCRIPTION
**Pull request by llogiq**
Btw. This is the first use of the *parent* logic (and I still want to add a compiling test for that).

Oh, and it closes #121

### CODE REVIEW

File: src/strings.rs [hunk 1]
```diff
@@ -17,11 +17,49 @@ declare_lint! {
...
+
+#[derive(Copy, Clone)]
 pub struct StringAdd;
 
 impl LintPass for StringAdd {
     fn get_lints(&self) -> LintArray {
+        lint_array!(STRING_ADD)
+    }
+
+    fn check_expr(&mut self, cx: &Context, e: &Expr) {
```
Comment (birkenfeld) [inline src/strings.rs]: would it make sense to put both lints in the same LintPass?

**Comment by birkenfeld** [comment]
This reintroduces mixed space-tab indentation...

File: src/strings.rs (hunk 1 above)
Comment (llogiq) [inline src/strings.rs]: I thought about it, but as most users will keep the `string_add` `allow`ed, I decided against it for now.

**Comment by llogiq** [comment]
Ouch. Fixed.

File: src/strings.rs (hunk 1 above)
Comment (birkenfeld) [inline src/strings.rs]: You can keep both STRING_ADD and STRING_ADD_ASSIGN, but they can both be emitted from the StringAdd LintPass (like in `methods.MethodsPass`).

It saves a bit of code and keeps related lints together, but mostly it's a style question I guess.

Comment (llogiq) [inline src/strings.rs]: I had thought that `allow`ed lints would not even be executed. That's why I put them apart.

Comment (birkenfeld) [inline src/strings.rs]: I think you can't disable passes, just individual lints.

Comment (Manishearth) [inline src/strings.rs]: Yeah, the disabling is done in individual lints, not passes. The passes are unconditionally all run simultaneously (It's a single AST visit)

File: src/lib.rs [hunk 2]
```diff
@@ -56,6 +56,7 @@ pub fn plugin_registrar(reg: &mut Registry) {
     reg.register_lint_pass(box misc::ModuloOne as LintPassObject);
     reg.register_lint_pass(box unicode::Unicode as LintPassObject);
     reg.register_lint_pass(box strings::StringAdd as LintPassObject);
+    reg.register_lint_pass(box strings::StringAddAssign as LintPassObject);
```
Comment (Manishearth) [inline src/lib.rs]: StringAdd needs to be in the lint group too

Comment (llogiq) [inline src/lib.rs]: But it's `allow`d by default. Wouldn't that make `#[warn(clippy)]` also warn on `string_add`?

Comment (Manishearth) [inline src/lib.rs]: Yes.

We can have multiple lint groups (`clippy`, `clippy_allow`, etc) if people want more fine grained support.

Comment (birkenfeld) [inline src/lib.rs]: (If something is done about that, there are two more lints that are `Allow` by default, `PTR_ARG` and `RESULT_UNWRAP_USED`.)

Comment (llogiq) [inline src/lib.rs]: OK then.

**Comment by llogiq** [comment]
@Manishearth r? :smile:

File: src/strings.rs [hunk 3]
```diff
@@ -17,16 +17,40 @@ declare_lint! {
...
+                    if let Some(ref p) = parent {
+                        if let &ExprAssign(ref target, _) = &p.node {
+                            // avoid duplicate matches
+                            if is_exp_equal(target, left) { return; }
+                        }
+                    }
+                }
+                //TODO check for duplicates
+                 span_lint(cx, STRING_ADD, e.span,
+                        "you add something to a string. \
```
Comment (Manishearth) [inline src/strings.rs]: *added

**Comment by Manishearth** [comment]
lgtm modulo grammar

It seems like a lot of folks don't find this to be problematic though. Perhaps both lints should be Allow?

**Comment by llogiq** [comment]
Ok, we can start allow by default. I have spent my anger at the core team's decision to implement Add for String (value-taking and asymmetric, no less!), and I probably should stop misusing clippy to try to convince others. :smile:

**Comment by Manishearth** [comment]
Btw, rebase instead of merging, it's harder to review :)

**Comment by llogiq** [comment]
OK. I'm still wrestling with git now and then (I'm used to SVN, so ...)

**Comment by Manishearth** [comment]
So if you want to sync your PR with the latest master, do:

```
git checkout <branch>
git fetch origin
git rebase -i branch
```

instruct it to pick all the commits (optionally, change "pick" to "fixup" for all commits you want squashed into the previous commits), and fix merge conflicts if they come up.

I'm also okay with manually merging, so it's fine to leave a PR with merge conflicts and I'll fix them when I merge.

If you just want to merge two trees, then use `git merge`, but one rarely needs to do that in a single PR

**Comment by llogiq** [comment]
Thanks!

**Pull request merged**